### PR TITLE
skip to replace link text if it is a link

### PR DESCRIPTION
**Pull request by eh2077**
### Fixed Issues
$ https://github.com/Expensify/App/issues/16475#issuecomment-1492798632

# Tests
1. Go to a chat and add a new comment, like `test`
2. Edit the comment to replace it with the following link and save it
```
https://www.facebook.com/hashtag/__main/?__eep__=6
```
3. Verify the link text and url are same as the original link
4. Edit the link to remove the link by replacing the link markdown the following original link and save it
```
https://www.facebook.com/hashtag/__main/?__eep__=6
```
5. Verify the link is not autolinked and displayed as text `https://www.facebook.com/hashtag/__main/?__eep__=6`

https://user-images.githubusercontent.com/117511920/232011057-cec36463-f327-4a1e-8b35-a097e3ef889e.mov

# QA
Same as Tests

### CODE REVIEW

**Comment by amyevans** [comment]
@fedirjh please review, GH is not showing me the option to assign you as a reviewer. Thanks!

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -402,12 +402,18 @@ export default class ExpensiMark {
...
             } else {
-                // `match[1]` contains the text inside []. Text is already escaped due to the rules
-                // that precede the link rule (eg, codeFence, inlineCodeBlock, email).
-                const linkText = this.replace(match[1], {
-                    filterRules: ['bold', 'strikethrough', 'italic'],
-                    shouldEscapeText: false,
-                });
+                let linkText = match[1];
+
+                // We don't want to apply filterRules if the linkText is entirely a link
```
Comment (marcaaron) [inline lib/ExpensiMark.js:407]: I think this comment would be more valuable if it explained **why** we don't want to "apply `filterRules`" and what that means exactly. It's also a little unclear what "entirely a link" means vs. something that is "partially a link" or some other meaning.

File: lib/ExpensiMark.js [hunk 2]
```diff
@@ -402,12 +402,18 @@ export default class ExpensiMark {
...
+                const urlRegex = new RegExp(`^${URL_REGEX}$`, 'gi');
+                if (!urlRegex.test(linkText)) {
+                    // `match[1]` contains the text inside []. Text is already escaped due to the rules
```
[the remark this answers is not in the document]
Comment (marcaaron) [reply lib/ExpensiMark.js:410]: >Text is already escaped due to the rules that precede the link rule (eg, codeFence, inlineCodeBlock, email).

This feels out of context or could be explained better. What would happen if we "escaped" the text twice?

**Review by marcaaron — CHANGES_REQUESTED** [review]
Code LGTM - had a few suggestions about the code comments.

File: __tests__/ExpensiMark-HTML-test.js [hunk 3]
```diff
@@ -366,7 +366,8 @@ test('Test url replacements', () => {
         + 'https://bastion1.sjc/logs/app/kibana#/discover?_g=()&_a=(columns:!(_source),index:\'2125cbe0-28a9-11e9-a79c-3de0157ed580\',interval:auto,query:(language:lucene,query:\'\'),sort:!(timestamp,desc)) '
         + 'google.com/maps/place/The+Flying\'+Saucer/@42.4043314,-86.2742418,15z/data=!4m5!3m4!1s0x0:0xe28f6108670216bc!8m2!3d42.4043316!4d-86.2742121 '
         + 'google.com/maps/place/%E9%9D%92%E5%B3%B6%E9%80%A3%E7%B5%A1%E8%88%B9%E4%B9%97%E5%A0%B4/@33.7363156,132.4877213,17.78z/data=!4m5!3m4!1s0x3545615c8c65bf7f:0xb89272c1a705a33f!8m2!3d33.7366776!4d132.4878843 '
-        + 'https://www.google.com/maps/place/Taj+Mahal+@is~"Awesome"/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422';
+        + 'https://www.google.com/maps/place/Taj+Mahal+@is~"Awesome"/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422 '
+        + 'https://www.facebook.com/hashtag/__main/?__eep__=6';
```
[the remark this answers is not in the document]
Comment (eh2077) [reply __tests__/ExpensiMark-HTML-test.js:370]: Sure, updated

File: lib/ExpensiMark.js [hunk 4]
```diff
@@ -402,12 +402,18 @@ export default class ExpensiMark {
...
+                    // that precede the link rule (eg, codeFence, inlineCodeBlock, email).
+                    linkText = this.replace(match[1], {
+                        filterRules: ['bold', 'strikethrough', 'italic'],
+                        shouldEscapeText: false,
+                    });
+                }
```
[the remark this answers is not in the document]
Comment (eh2077) [reply lib/ExpensiMark.js:416]: Updated

File: lib/ExpensiMark.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (eh2077) [reply lib/ExpensiMark.js:410]: Updated

File: __tests__/ExpensiMark-HTML-test.js [hunk 5]
```diff
@@ -399,7 +402,10 @@ test('Test url replacements', () => {
...
+        + '<a href="https://www.google.com/maps/place/Taj+Mahal+@is~&quot;Awesome&quot;/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422" target="_blank" rel="noreferrer noopener">https://www.google.com/maps/place/Taj+Mahal+@is~&quot;Awesome&quot;/@27.1751496,78.0399535,17z/data=!4m12!1m6!3m5!1s0x39747121d702ff6d:0xdd2ae4803f767dde!2sTaj+Mahal!8m2!3d27.1751448!4d78.0421422!3m4!1s0x39747121d702ff6d:0xdd2ae4803f767dde!8m2!3d27.1751448!4d78.0421422</a> '
+        + '<a href="https://www.facebook.com/hashtag/__main/?__eep__=6" target="_blank" rel="noreferrer noopener">https://www.facebook.com/hashtag/__main/?__eep__=6</a> '
+        + '<a href="https://example.com/~username/foo~bar.txt" target="_blank" rel="noreferrer noopener">https://example.com/~username/foo~bar.txt</a> '
+        + '<a href="http://example.com/foo/*/bar/*/test.txt" target="_blank" rel="noreferrer noopener">http://example.com/foo/*/bar/*/test.txt</a>';
```
Comment (amyevans) [inline __tests__/ExpensiMark-HTML-test.js:408]: Hmm can you help me understand why we expect `http://example.com/foo/*/bar/*/test.txt` to be identified as and converted to a link? I'm reading [this article](https://perishablepress.com/stop-using-unsafe-characters-in-urls/#:~:text=These%20characters%20are%20%7B%20%2C%20%7D%20%2C,be%20encoded%20within%20a%20URL.) which references the [official spec](https://www.rfc-editor.org/rfc/rfc3986) and my takeaway is that `*` is a reserved character and would need to be URL-encoded in order to be valid, so I'm not sure we should have this as a test case (but I could also be misunderstanding something).

**Review by amyevans — COMMENTED** [review]
One question, but otherwise LGTM

File: __tests__/ExpensiMark-HTML-test.js (hunk 5 above)
Comment (eh2077) [reply __tests__/ExpensiMark-HTML-test.js:408]: Hi @amyevans , that's a great question! 

The article you linked says that

> 1 Reserved characters only need encoded when not used for their defined, reserved purposes.

When I tried to add the case of `http://example.com/foo/*/bar/*/test.txt`, I also found that the asterisk character is not alway valid in a url but has to be used in its defined and reserved purpose way. This test case url `http://example.com/foo/*/bar/*/test.txt` should follow the defined purpose. I checked it with the [web url regex](https://github.com/Expensify/expensify-common/blob/be625f55058111793bbcfe3f64788ea0a8f9d705/lib/Url.js#L8) provided by this repo, see

image

You may test more cases about link includes the asterisk characters and below the web url regex could be helpful 
```js
/((https?:\/\/)?((?:www\.)?[-a-z0-9]+?\.)+(?:COM|ORG)(?:\:\d{2,4}|\b|(?=_))(?:(?:[.,=(+$!*]|&(?:amp|quot|#x27);)?\/(?:[-\w$@.+!*:(),=%~]|&(?:amp|quot|#x27);)*(?:[-\w~@:%)]|&(?:amp|quot|#x27);)|\/)*(?:(?:\?(?:[-\w$@.+!*()\/,=%{}:;\[\]\|_]|&(?:amp|quot|#x27);)*)?|(?:#(?:[-\w$@.+!*()[\],=%;\/:~]|&(?:amp|quot|#x27);)*)?)*)/
```

Comment (amyevans) [reply __tests__/ExpensiMark-HTML-test.js:408]: Makes sense, thank you both for the additional discussion!

**Review by amyevans — APPROVED** [review]

**Comment by amyevans** [comment]
@marcaaron over to you for re-review 😄

**Review by marcaaron — APPROVED** [review]
Looks great, thanks for all the comment improvements!

**Pull request merged**